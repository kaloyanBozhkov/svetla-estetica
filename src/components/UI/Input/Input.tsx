import type { ChangeEvent, ReactElement } from 'react'
import styles from './input.module.scss'

export type InputProps = {
    // TODO setup other input types?
    type?: 'input' | 'textarea'
    name: string
    label: string
    value?: string
    required?: boolean
    description?: string
    modifier?: 'base' | 'baseActive'
    onChange?: (event: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => void
    onFocus?: (event: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => void
    onBlur?: (event: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => void
}

const Input = ({
    type = 'input',
    label,
    value = '',
    onChange,
    onFocus,
    onBlur,
}: InputProps): ReactElement => (
    <div className={styles.input} data-is-active={!!value} data-is-textarea={type === 'textarea'}>
        {type === 'textarea' ? (
            <textarea
                id={label}
                value={value}
                onChange={onChange}
                onFocus={onFocus}
                onBlur={onBlur}
            />
        ) : (
            <input id={label} value={value} onChange={onChange} onFocus={onFocus} onBlur={onBlur} />
        )}
        <label htmlFor={`${label}`}>{label}</label>
    </div>
)

export default Input
