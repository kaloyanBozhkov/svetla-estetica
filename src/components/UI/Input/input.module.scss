@import "scss/variables.scss";
@mixin activeLabel {
    top: 0;
    transform: translateY(-50%);
    font-size: 1.3rem;
    opacity: 1;
    padding: 0 0.5rem;
}

.input {
    width: 100%;
    height: 100%;
    position: relative;

    label {
        align-items: center;
        transition: 200ms;
        cursor: text;
        font-size: 1.6rem;
        left: 1rem;
        position: absolute;
        background-color: white;
        padding: 0;
        line-height: 1.6rem;
        z-index: $zIndexMax;
    }

    &[data-is-active="false"] {
        label {
            top: 50%;
            transform: translateY(-50%);
            opacity: 0.5;
        }
    }

    &[data-is-active="true"] label {
        @include activeLabel();
    }

    input,
    textarea {
        width: 100%;
        height: 100%;
        padding: 1rem;
        font-size: 1.6rem;
        border: 1px solid $lighterGrey;
        outline: none;
        flex: 1;
        border-radius: 4px;
        resize: none;
        z-index: $zIndexTop;

        &:-webkit-autofill {
            -webkit-box-shadow: 0 0 0 1000px white inset;
        }

        &:focus {
            border-color: $pink;
            ~ label {
                color: $pink;

                @include activeLabel();
            }
        }
    }

    &[data-is-textarea="true"] {
        &[data-is-active="false"] label {
            top: 2.1rem;
            opacity: 0.5;
        }

        &[data-is-active="true"] label {
            @include activeLabel();
        }
    }
}
