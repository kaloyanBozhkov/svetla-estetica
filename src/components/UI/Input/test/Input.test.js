import { shallow } from 'enzyme'
import Input from '../Input'

describe('testing Input atom', () => {
    it('should render atom', () => {
        expect(shallow(<Input value="" label="name" onChange={jest.fn()} />)).toMatchSnapshot()
    })
    it('should run onChange fn', () => {
        const mockOnChange = jest.fn(),
            wrapper = shallow(<Input value="" label="name" onChange={mockOnChange} />)
        wrapper.find('input').prop('onChange')()
        expect(mockOnChange).toHaveBeenCalled()
    })

    it('should render with additional modifier attribute if value set', () => {
        const wrapper = shallow(<Input value="some value" label="name" onChange={jest.fn()} />)
        expect(wrapper.prop('data-is-active')).toBe(true)
    })

    it('should render with textarea', () => {
        const wrapper = shallow(
            <Input type="textarea" value="" label="text" onChange={jest.fn()} />
        )
        expect(wrapper.find('textarea').length).toBe(1)
    })
})
