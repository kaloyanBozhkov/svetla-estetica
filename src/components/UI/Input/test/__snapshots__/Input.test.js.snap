// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing Input atom should render atom 1`] = `
<div
  className="input"
  data-is-active={false}
  data-is-textarea={false}
>
  <input
    id="name"
    onChange={[MockFunction]}
    value=""
  />
  <label
    htmlFor="name"
  >
    name
  </label>
</div>
`;
